src/extlibs/hrvo/simulator.cpp

```diff
@@ -36,96 +36,243 @@
+                     static_cast<float>(robot.velocity().y()));
+    float max_accel = 0.f;
+    float max_speed = robot.robotConstants().robot_max_speed_m_per_s;
+
+    // Max distance which the robot can travel in one time step + scaling
+    float goal_radius = (max_speed * timeStep_) / 2 * goal_radius_scale;
+
+    // Enemy agents should appear larger to friendly agents to avoid collision
+    float agent_radius = ROBOT_MAX_RADIUS_METERS * enemy_robot_radius_scale;
+
+    return addLinearVelocityAgent(position, agent_radius, velocity, max_speed, max_accel,
+                                  addGoal(destination), goal_radius);
+}
+
+std::size_t Simulator::addHRVOAgent(const Vector2 &position, float agent_radius,
+                                    const Vector2 &curr_velocity, float maxSpeed,
+                                    float prefSpeed, float maxAccel,
+                                    std::size_t goal_index, float goalRadius,
+                                    float neighborDist, std::size_t maxNeighbors,
+                                    float uncertaintyOffset)
+{
+    std::unique_ptr<HRVOAgent> agent = std::make_unique<HRVOAgent>(
+        this, position, goal_index, neighborDist, maxNeighbors, agent_radius,
+        curr_velocity, maxAccel, goalRadius, prefSpeed, maxSpeed, uncertaintyOffset);
+    agents_.push_back(std::move(agent));
     return agents_.size() - 1;
 }
 
-std::size_t Simulator::addAgent(const Vector2 &position, std::size_t goalNo,
-                                float neighborDist, std::size_t maxNeighbors,
-                                float radius, float goalRadius, float prefSpeed,
-                                float maxSpeed, float uncertaintyOffset, float maxAccel,
-                                const Vector2 &velocity, float orientation)
+size_t Simulator::addLinearVelocityAgent(const Vector2 &position, float agent_radius,
+                                         const Vector2 &curr_velocity, float max_speed,
+                                         float max_accel, size_t goal_index,
+                                         float goal_radius)
 {
-    Agent *const agent = new Agent(this, position, goalNo, neighborDist, maxNeighbors,
-                                   radius, velocity, maxAccel, goalRadius, prefSpeed,
-                                   maxSpeed, orientation, uncertaintyOffset);
-    agents_.push_back(agent);
+    std::unique_ptr<LinearVelocityAgent> agent = std::make_unique<LinearVelocityAgent>(
+        this, position, agent_radius, curr_velocity, max_speed, max_accel, goal_index,
+        goal_radius);
 
+    agents_.push_back(std::move(agent));
     return agents_.size() - 1;
 }
 
 std::size_t Simulator::addGoal(const Vector2 &position)
 {
-    Goal *const goal = new Goal(position);
-    goals_.push_back(goal);
+    std::unique_ptr<Goal> goal = std::make_unique<Goal>(position);
+    goals_.push_back(std::move(goal));
```

Comment:
I'm confused by this, shouldn't goals be on a per agent basis?

Reply:
Yes it should. The original HRVO implementation had it implemented this way, where Agent and Goal are both owned by the Simulator, rather than each Agent owning its own Goal. This is being fixed by @JoshuaWu7 in #2373